File: cpp/base/src/main/vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidLoop.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidFace.h"
#include "vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidHalfEdge.h"
#include "vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidLoop.h"
#include "vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidVertex.h"
_PolyhedralBoundedSolidLoop::_PolyhedralBoundedSolidLoop(_PolyhedralBoundedSolidFace* parent)
    : parentFace(parent), boundaryStartHalfEdge(0)
{
    if ( parentFace != 0 ) {
        parentFace->boundariesList.add(this);
    }
}

_PolyhedralBoundedSolidLoop::~_PolyhedralBoundedSolidLoop()
{
    for (long int i = 0; i < halfEdgesList.size(); ++i) {
        if (halfEdgesList[i] != 0) {
            delete halfEdgesList[i];
        }
    }
}

void _PolyhedralBoundedSolidLoop::unlistHalfEdge(_PolyhedralBoundedSolidHalfEdge* he)
{
    halfEdgesList.remove(he);
    boundaryStartHalfEdge = halfEdgesList.size() == 0 ? 0 : halfEdgesList[0];
}

_PolyhedralBoundedSolidHalfEdge* _PolyhedralBoundedSolidLoop::halfEdgeVertices(int a, int b)
{
    if ( halfEdgesList.size() == 0 ) return 0;
    for (long int i = 0; i < halfEdgesList.size(); ++i) {
        _PolyhedralBoundedSolidHalfEdge* oldhe = halfEdgesList[i];
        _PolyhedralBoundedSolidHalfEdge* he = oldhe->next();
        if ( he != 0 && oldhe->startingVertex != 0 && he->startingVertex != 0 &&
             oldhe->startingVertex->id == a && he->startingVertex->id == b ) {
            return oldhe;
        }
    }
    return 0;
}

_PolyhedralBoundedSolidHalfEdge* _PolyhedralBoundedSolidLoop::firstHalfEdgeAtVertex(int a)
{
    for (long int i = 0; i < halfEdgesList.size(); ++i) {
        _PolyhedralBoundedSolidHalfEdge* oldhe = halfEdgesList[i];
        if ( oldhe->startingVertex != 0 && oldhe->startingVertex->id == a ) {
            return oldhe;
        }
    }
    return 0;
}

void _PolyhedralBoundedSolidLoop::delhe(_PolyhedralBoundedSolidHalfEdge* he)
{
    unlistHalfEdge(he);
}

void _PolyhedralBoundedSolidLoop::revert()
{
    long int n = halfEdgesList.size();
    for (long int i = 0; i < n / 2; i++) {
        _PolyhedralBoundedSolidHalfEdge* tmp = halfEdgesList[i];
        halfEdgesList[i] = halfEdgesList[n - 1 - i];
        halfEdgesList[n - 1 - i] = tmp;
    }
    boundaryStartHalfEdge = n == 0 ? 0 : halfEdgesList[0];
}

_PolyhedralBoundedSolidHalfEdge* _PolyhedralBoundedSolidLoop::previousOf(_PolyhedralBoundedSolidHalfEdge* he) const
{
    long int n = halfEdgesList.size();
    if ( he == 0 || n == 0 ) return 0;
    for (long int i = 0; i < n; ++i) {
        if ( halfEdgesList.get(i) == he ) {
            long int j = (i == 0) ? n - 1 : i - 1;
            return halfEdgesList.get(j);
        }
    }
    return 0;
}

_PolyhedralBoundedSolidHalfEdge* _PolyhedralBoundedSolidLoop::nextOf(_PolyhedralBoundedSolidHalfEdge* he) const
{
    long int n = halfEdgesList.size();
    if ( he == 0 || n == 0 ) return 0;
    for (long int i = 0; i < n; ++i) {
        if ( halfEdgesList.get(i) == he ) {
            return halfEdgesList.get((i + 1) % n);
        }
    }
    return 0;
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidLoop.ts
//= References:                                                             =
//= [MANT1988] Mantyla Martti. "An Introduction To Solid Modeling",         =
//=     Computer Science Press, 1988.                                       =

import { CircularDoubleLinkedList } from "../../../../../common/dataStructures/CircularDoubleLinkedList.js";
import { FundamentalEntity } from "../../../../../common/FundamentalEntity.js";
import type { _PolyhedralBoundedSolidEdge } from "./_PolyhedralBoundedSolidEdge.js";
import type { _PolyhedralBoundedSolidFace } from "./_PolyhedralBoundedSolidFace.js";
import type { _PolyhedralBoundedSolidHalfEdge } from "./_PolyhedralBoundedSolidHalfEdge.js";

/**
As noted in [MANT1988].10.2.1, a `_PolyhedralBoundedSolidLoop` describes
one connected boundary inside a `_PolyhedralBoundedSolidFace`.

Note that in the sake of simplify and eficiency current programming
implementation of this class exhibit public access attributes. It is important
to note that those attributes will only be accessed directly from related
classes in the same package
(vsdk.toolkit.environment.geometry.volume.polyhedralBoundedSolid.nodes) and
from methods in the `PolyhedralBoundedSolid` class, and that they should
not be used from outer classes.
*/
export class _PolyhedralBoundedSolidLoop extends FundamentalEntity {
    /// Defined as presented in [MANT1988].10.2.1
    public parentFace!: _PolyhedralBoundedSolidFace;

    /// As noted in [MANT1988].10.2.3, consider that there is a special
    /// case for empty loops. Note that this case doesn't affect this
    /// reference.
    public boundaryStartHalfEdge: _PolyhedralBoundedSolidHalfEdge | null = null;

    public halfEdgesList!: CircularDoubleLinkedList<_PolyhedralBoundedSolidHalfEdge>;

    //=================================================================
    public constructor(parent: _PolyhedralBoundedSolidFace) {
        super();
        this.init(parent);
    }

    private init(parent: _PolyhedralBoundedSolidFace): void {
        this.parentFace = parent;
        this.parentFace.boundariesList.add(this);
        this.halfEdgesList = new CircularDoubleLinkedList<_PolyhedralBoundedSolidHalfEdge>();
    }

    public unlistHalfEdge(he: _PolyhedralBoundedSolidHalfEdge): void {
        if (this.halfEdgesList.locateWindowAtElem(he)) {
            this.halfEdgesList.removeElemAtWindow();
        }
        if (this.halfEdgesList.size() > 0) {
            this.boundaryStartHalfEdge = this.halfEdgesList.get(0);
        } else {
            this.boundaryStartHalfEdge = null;
        }
    }

    /**
    Locates a half edge that goes from vertex with id `a` to vertex with
    id `b`.  Returns null if no such half edge exists in this loop.
    @param a
    @param b
    @return requested halfedge
    */
    public halfEdgeVertices(a: number, b: number): _PolyhedralBoundedSolidHalfEdge | null {
        let he: _PolyhedralBoundedSolidHalfEdge | null;
        let oldhe: _PolyhedralBoundedSolidHalfEdge;
        he = this.boundaryStartHalfEdge;
        do {
            oldhe = he!;
            he = he!.next();
            if (he === null) {
                // Loop is not closed!
                break;
            }

            if (oldhe.startingVertex.id === a && he.startingVertex.id === b) {
                return oldhe;
            }
        } while (he !== this.boundaryStartHalfEdge);
        return null;
    }

    /** Locates a half edge that goes from vertex with id `a` to vertex with
    id `b`.  Returns null if no such half edge exists in this loop. */
    public firstHalfEdgeAtVertex(a: number): _PolyhedralBoundedSolidHalfEdge | null {
        let he: _PolyhedralBoundedSolidHalfEdge | null;
        let oldhe: _PolyhedralBoundedSolidHalfEdge;
        he = this.boundaryStartHalfEdge;
        do {
            oldhe = he!;
            he = he!.next();
            if (he === null) {
                // Loop is not closed!
                break;
            }

            if (oldhe.startingVertex.id === a) {
                return oldhe;
            }
        } while (he !== this.boundaryStartHalfEdge);
        return null;
    }

    /**
    Vitral SDK's current implementation of original `delhe` utility function
    presented at program [MANT1988].11.4. and section [MANT1988].11.2.2.
    Note that current implementation is quite diferent from the original from
    [MANT1988]. This could lead to subtle problems! This method's functionality
    should be better understood!
    */
    public delhe(he: _PolyhedralBoundedSolidHalfEdge): void {
        this.halfEdgesList.locateWindowAtElem(he);
        this.halfEdgesList.removeElemAtWindow();

        if (this.halfEdgesList.size() > 0) {
            this.boundaryStartHalfEdge = this.halfEdgesList.get(0);
        }
    }

    public revert(): void {
        if (this.halfEdgesList.size() <= 0) {
            return;
        }

        //-----------------------------------------------------------------
        let he: _PolyhedralBoundedSolidHalfEdge;
        const edges: (_PolyhedralBoundedSolidEdge | null)[] = new Array(this.halfEdgesList.size()).fill(null);
        const sides: boolean[] = new Array(this.halfEdgesList.size()).fill(false);
        let i: number;

        for (i = 0; i < this.halfEdgesList.size(); i++) {
            he = this.halfEdgesList.get(i)!;
            // Degenerate loops can be left behind transiently after lkemr
            // during boolean finishing. They are not traversable via an edge.
            if (he.parentEdge === null) {
                return;
            }
            edges[i] = he.parentEdge;
            sides[i] = false;
            if (he.parentEdge.rightHalf === he) {
                sides[i] = true;
            }
        }

        for (i = 1; i < this.halfEdgesList.size(); i++) {
            this.halfEdgesList.get(i)!.parentEdge = edges[i - 1]!;
            if (sides[i - 1]) {
                edges[i - 1]!.rightHalf = this.halfEdgesList.get(i);
            } else {
                edges[i - 1]!.leftHalf = this.halfEdgesList.get(i);
            }
        }
        this.halfEdgesList.get(0)!.parentEdge = edges[i - 1]!;
        if (sides[i - 1]) {
            edges[i - 1]!.rightHalf = this.halfEdgesList.get(0);
        } else {
            edges[i - 1]!.leftHalf = this.halfEdgesList.get(0);
        }

        //-----------------------------------------------------------------
        this.halfEdgesList.reverse();
    }

    public override toString(): string {
        let msg: string;

        msg = "Loop, parent face " + this.parentFace.id;

        return msg;
    }
}
